cpp/dolfinx/mesh/MeshTags.h

```diff
@@ -335,28 +333,33 @@ create_meshtags(MPI_Comm comm, const std::shared_ptr<const mesh::Mesh>& mesh,
     entities_igi.insert({entity_igi, e});
   }
 
+  const graph::AdjacencyList<std::int64_t> recv_ents
+      = MPI::all_to_all(comm, graph::AdjacencyList<std::int64_t>(send_ents));
+  const graph::AdjacencyList<T> recv_vals
+      = MPI::all_to_all(comm, graph::AdjacencyList<T>(send_vals));
+
   // Iterate over all received entities and find it in entities of the
   // mesh
 
   // FIXME: Avoid creating std::vector<std::int64_t> the loop below
 
   std::vector<std::int32_t> indices_new;
   std::vector<T> values_new;
-  for (std::int32_t i = 0; i < comm_size; ++i)
+  for (std::int32_t i = 0; i < recv_ents.num_nodes(); ++i)
   {
-    const std::int32_t num_recv_ents
-        = (std::int32_t)(recv_ents[i].size() / nnodes_per_entity);
-    for (std::int32_t e = 0; e < num_recv_ents; ++e)
+    auto recv_ents_p = recv_ents.links(i);
+    auto recv_vals_p = recv_vals.links(i);
+    for (std::int32_t e = 0; e < recv_ents_p.rows() / nnodes_per_entity; ++e)
     {
-      std::vector<std::int64_t> _entity(&recv_ents[i][nnodes_per_entity * e],
-                                        &recv_ents[i][nnodes_per_entity * e]
-                                            + nnodes_per_entity);
+      std::vector<std::int64_t> _entity(
+          recv_ents_p.data() + nnodes_per_entity * e,
```

Comment:
I think `recv_ents_p[e].data()` and `recv_ents_p[e].data() + nnodes_per_entity` could work here.

Reply:
Code block now substantially re-structured.